verilog/preprocessor/verilog_preprocess.cc

```diff
@@ -552,6 +554,115 @@ absl::Status VerilogPreprocess::HandleEndif(
   return absl::OkStatus();
 }
 
+// Handle `include directives.
+absl::Status VerilogPreprocess::HandleInclude(
+    TokenStreamView::const_iterator iter,
+    const StreamIteratorGenerator& generator) {
+  // TODO(karimtera): how to differentiate between <file> and "file"?? both are
+  // the same token, need to edit the lexer.
+  TokenStreamView::const_iterator token_iter = generator();
+  auto file_token_iter = *token_iter;
+  if (file_token_iter->token_enum() != TK_StringLiteral) {
+    preprocess_data_.errors.push_back(
+        {**token_iter, "Expected a path to a SV file."});
+    return absl::InvalidArgumentError("Expected a path to a SV file.");
+  }
+  // Currently the file path looks like "path", we need to remove "".
+  const auto& token_text = file_token_iter->text();
+
+  std::filesystem::path file_path =
+      std::string(token_text.substr(1, token_text.size() - 2));
+
+  std::string file_path_to_search = file_path.string();
+
+  if (file_path.is_relative()) {
+    bool found = false;
+    for (const auto& base_dir : preprocess_info_.include_dirs) {
+      file_path_to_search =
+          verible::file::JoinPath(base_dir, file_path.string());
+      auto search_status = verible::file::FileExists(file_path_to_search);
+      if (search_status.ok()) {
+        found = true;
+        break;
+      }
+    }
+    if (!found) {
+      preprocess_data_.errors.push_back({**token_iter, "no such file found."});
+      return absl::InvalidArgumentError("ERROR: included file can't be found.");
+    }
+  } else {
+    if (auto status = verible::file::FileExists(file_path_to_search);
+        !status.ok()) {
+      preprocess_data_.errors.push_back({**token_iter, "no such file found."});
+      return absl::InvalidArgumentError("ERROR: included file can't be found.");
+    }
+  }
+
+  // Actually open the file.
+  std::string source_contents;
+  if (auto status =
+          verible::file::GetContents(file_path_to_search, &source_contents);
+      !status.ok()) {
+    preprocess_data_.errors.push_back(
+        {**token_iter, "ERROR: passed file can't be open"});
+    return absl::InvalidArgumentError("ERROR: passed file can't be open.");
+  }
+
+  // Creating a new "VerilogPreprocess" object for the included file,
+  // With the same configuration and preprocessing info (defines, incdirs) as
+  // the main one.
+  verilog::VerilogPreprocess child_preprocessor(config_);
+  child_preprocessor.setPreprocessingInfo(preprocess_info_);
+
+  // TODO(karimtera): limit number of nested includes, detect cycles? maybe.
+  preprocess_data_.included_text_structure.emplace_back(
```

Comment:
Is it planned to have a map of filename -> file content at some point, so that if we encounter the same file multiple times, we only have it in memory once ?

`class VerilogProject` might already do some of this, but if this does not match right now, add a TODO, to merge that functionality (there is the `file_set_type files_` member that contains that info, also used to avoid re-loading the same file multiple times).

Reply:
I looked at it again, I think it can be done, But we will need to change the public API provided by `VerilogPreprocess`, because right now `ScanStream(TokenSequence)` is the only one available, I will add it as a TODO.